File: java/MaximumDepthOfBinaryTree.java
// public class MaximumDepthOfBinaryTree {
//   TreeNode sampleNode = new TreeNode(
//     3,
//     new TreeNode(9),
//     new TreeNode(20, 
//       new TreeNode(15),
//       new TreeNode(7))
//     );
//   public static void main(String[] args) {
//     System.out.println(maxDepth(sampleNode));
//   }

//   static final int SEARCHED = 1000;
//   public int maxDepth(TreeNode root) {
//     if (root == null) return 0;
//     return dfs(root, 1);
//   }

//   public int dfs(TreeNode node, int count) {
//     if (node == null) {
//       return count;
//     }

//     node.val = SEARCHED;
//     var l = node.left;
//     var r = node.right;
    
//     if (isSearched(l) && isSearched(r)) {
//       return count;
//     } else if(isSearched(l) && !isSearched(r)) {
//       return dfs(r, ++count);
//     } else if (!isSearched(l) && isSearched(r)) {
//       return dfs(l, ++count);
//     } else {
//       // 左右どちらも未探索の場合は一旦左を探索
//       return dfs(l, ++count);
//     }
//   }

//   public boolean isSearched(TreeNode node) {
//     return node == null || node.val == SEARCHED;
//   }

//   /**
//    * Definition for a binary tree node.
//    */
//   public class TreeNode {
//     int val;
//     TreeNode left;
//     TreeNode right;

//     TreeNode() {
//     }

//     TreeNode(int val) {
//       this.val = val;
//     }

//     TreeNode(int val, TreeNode left, TreeNode right) {
//       this.val = val;
//       this.left = left;
//       this.right = right;
//     }
//   }

// }

File: python/maximum_depth_of_binary_tree.py
# Definition for a binary tree node.
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        if root is None:
            return 0
        maxLeft = self.maxDepth(root.left)
        maxRight = self.maxDepth(root.right)
        return max(maxLeft, maxRight) + 1

s = Solution()
node = TreeNode(3, TreeNode(9), TreeNode(20, TreeNode(15), TreeNode(7)))
print(s.maxDepth(node))